File: src/mvc/cpp/string_list_model/main_window.cpp
#include "main_window.h"
#include "insert_element_dialog.h"
#include "./ui_main_window.h"
#include <QStringListModel>
#include <QModelIndex>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent), ui(new Ui::MainWindow) {
  ui->setupUi(this);

  auto model = new QStringListModel(this);
  model->setStringList(QStringList() << "Item 1" << "Item 2" << "Item 3");
    ui->listView->setModel(model);

    ui->listView->setEditTriggers(QAbstractItemView::DoubleClicked | QAbstractItemView::SelectedClicked);
    ui->listView->setSelectionMode(QAbstractItemView::MultiSelection);
    ui->listView->setDragEnabled(true);
    ui->listView->setAcceptDrops(true);
    ui->listView->setDropIndicatorShown(true);
    ui->listView->setDragDropMode(QAbstractItemView::InternalMove);

    connect(ui->insertButton, &QPushButton::clicked, this, [=]() {
        auto dialog = new InsertElementDialog(this);
        dialog->setMaxIndex(model->rowCount());
        connect(dialog, &InsertElementDialog::elementSelected, this, [=](InsertElementData data) {
            model->insertRows(data.index, 1);
            model->setData(model->index(data.index), data.text);
        });
        dialog->show();
    });

    connect(ui->removeButton, &QPushButton::clicked, this, [=]() {
        auto indexes = ui->listView->selectionModel()->selectedIndexes();
        for (auto index : indexes) {
            model->removeRow(index.row());
        }
    });

}

MainWindow::~MainWindow() { delete ui; }

File: src/mvc/python/string_list_model/main_window.py
from PyQt6 import QtWidgets, uic
from PyQt6.QtCore import QDir, QStringListModel
from PyQt6.QtWidgets import QAbstractItemView

from insert_element_dialog import InsertElementDialog


class MainWindow(QtWidgets.QMainWindow):
    def __init__(self):
        super().__init__()
        uic.loadUi("main_window.ui", self)

        model = QStringListModel(self)
        model.setStringList(["Item 1", "Item 2", "Item 3"])
        self.listView.setModel(model)

        self.listView.setEditTriggers(
            QAbstractItemView.EditTrigger.DoubleClicked
            | QAbstractItemView.EditTrigger.SelectedClicked
        )

        self.listView.setSelectionMode(QAbstractItemView.SelectionMode.MultiSelection)
        self.listView.setDragEnabled(True)
        self.listView.setAcceptDrops(True)
        self.listView.setDropIndicatorShown(True)
        self.listView.setDragDropMode(QAbstractItemView.DragDropMode.InternalMove)

        self.insertButton.clicked.connect(self.insert_element)
        self.removeButton.clicked.connect(self.remove_element)
        self.show()

    def insert_element(self):
        dialog = InsertElementDialog(self)
        dialog.set_max_index(self.listView.model().rowCount())
        dialog.element_selected.connect(self.insert_element_data)
        dialog.show()

    def insert_element_data(self, data):
        self.listView.model().insertRows(data.index, 1)
        self.listView.model().setData(
            self.listView.model().index(data.index), data.text
        )

    def remove_element(self):
        indexes = self.listView.selectionModel().selectedIndexes()
        for index in indexes:
            self.listView.model().removeRow(index.row())
